rfc: Implement generators for XML text and attribute

tienvx opened the pull request.

YOU54F commented:
Associated draft pull requests

- https://github.com/pact-foundation/pact-reference/pull/486
- https://github.com/pact-foundation/pact-reference/pull/487
- https://github.com/pact-foundation/pact-php/pull/720

YOU54F commented:
Tagging @pact-foundation/maintainers for review of this RFC to introduce generators for XML text and attributes. Thank you 🙌🏾

rholshausen commented:
Just a note, the reason that generators are not supported for XML is because it is more complex than with JSON. The current implementation is based on finding the element in the body that matches the path, and replacing it with a generated value.

XML can have multiple matching nodes, and that would require being able to find all of them and replacing all the items. This is where I gave up on it.

I.e. with
```xml
<container>
  <tag/><tag/><tag/>
</container>
```
and the path `$.container.tag` would require 3 updates.

with
```xml
<container>
  <tag>
    <child>
        <subchild>
        </subchild>
        <subchild>
        </subchild>
    </child>
  <tag/>
  <tag>
  <tag/>
  <other>
     <child>
        <name>
        </name>
        <name>
        </name>
    </child>
  <other/>
</container>
```
and the path `$.*.child.*` would match 4 nodes (and think how this can expand on non-trivial documents).

rholshausen commented:
> I do have some questions and comments, and these mostly revolve around the complexity of XML as a format. For reference, I'm referring to the [XML 1.1 specification](https://www.w3.org/TR/2006/REC-xml11-20060816/).
> ## Support for XML Declaration
> ## Empty tags
> ## Support for white-space preserving
> ## Support for escaping data
> ## Support for Comments (?)
> ## Support for Type Declarations
> ## Support for Attribute-list Declarations
> ## Support for Entities

These are not relevant for using generators with XML.

JP-Ellis commented:
> These are not relevant for using generators with XML.

Care to go into any level of detail as to why?

tienvx commented:
> with
> ```
> <container>
>   <tag/><tag/><tag/>
> </container>
> ```
> and the path $.container.tag would require 3 updates.

This case has been handled in my first commit

> with
> ```
> <container>
>   <tag>
>     <child>
>         <subchild>
>         </subchild>
>         <subchild>
>         </subchild>
>     </child>
>   <tag/>
>   <tag>
>   <tag/>
>   <other>
>      <child>
>         <name>
>         </name>
>         <name>
>         </name>
>     </child>
>   <other/>
> </container>
> ```
> and the path $..container.*.child.* would match 4 nodes

Thanks for suggesting this case. I added more tests and update code to handle this case.

tienvx commented:
I updated the RFC to answer most of the questions. I also updated code to fix the multiple element in different path handling, and add more tests to check more edge cases.

tienvx commented:
This PR is ready for review again. Please take a look again!

tienvx commented:
It has been a while. I'd like your feedback on this PR again. @JP-Ellis @rholshausen @mefellows  . If everything is fine, an approval would be nice.
